### DISCUSSION TITLE
Heterogenous Sampling Rates Player

### DISCUSSION THREAD

**Discussion by GuillaumeLeclerc** [pschatzmann/arduino-audio-tools#550] [Q&A] [answered]
Hello,

I'm playing with the `Player` example with multiple web radios. However some are 44.1kHZ and some are 48KHz. What is the best way to handle this ? as I need a fixed sampling rate for the I2S output. I tried changing the I2S config but it get overwritten.

Thank you for the help!

**Comment by pschatzmann** [comment, marked as answer]
I am not sure what exactly your issue is: It is my expectation that the decoder is notifying the output about any changes in the sampling rate. If you set the log level to info you should see the related setAudioInfo() calls when e.g. the sampling rate or number of channels is changing...

**Reply by GuillaumeLeclerc** [reply]
It is changing indeed, but I can't afford to. The I2S has to have a fixed sampling rate and MCLK. It can't be updated when a button is pressed.

What I need is something that checks if the sampling rate is different than 48KHz it is resampled before being passed to the I2S stream.

**Comment by pschatzmann** [comment]
I suggest that you wrap the I2SStream in a ResampleStream. That should theoretically take care of your problem. However I never tested this scenario and in the worst case you will need to make a couple of corrections...

**Reply by GuillaumeLeclerc** [reply]
Thanks for the help!

However ResampleStream has a step_size argument not a "target_sampling_rate". Am I missing something ?

**Reply by pschatzmann** [reply]
Yes, that's the correction which would be necessary: I had a different (but less flexible) implementation before which has been replaced with the dynmaic float step size and it seems that I forgot to add that functionality: So the only thing that would be needed is to recalculate the step size when setAudioInfo() is called. See getStepSize(fromRate, toRate) method. 

I guess it might be necessary to add a new attribute: target_sample_rate to the ResampleConfig

**Reply by GuillaumeLeclerc** [reply]
Weirdly I fail at allocation on the buffer for resampling. My application is just the example web plaer. Is it expected to run out of RAM here ?

Thanks

**Reply by GuillaumeLeclerc** [reply]
I think the simplest for me will just to adjust the MCLK so that it's a sampling_rate * 256. This will mess up a little my DSP filters but I guess I'll have to live with that.

**Comment by pschatzmann** [comment]
Strange: if you just write data no additinal buffers should be allocated. I  need some additional buffers when you would call readBytes...

**Reply by GuillaumeLeclerc** [reply]
It is allocating two buffers, one for upSampling and one for down sampling. They are about 50k elements. The one for upSampling succeed but the second malloc fails.

**Reply by pschatzmann** [reply]
Can it be that you still use the old version of the library ? 
I suggest that you make a quick compare with the implementation on Github. The new optimized Resample.h should have only 201 lines of code: see https://www.pschatzmann.ch/home/2022/12/13/arduino-audio-tools-resampling-revisited/

**Reply by GuillaumeLeclerc** [reply]
That must be it then! I have more than 500. I installed the library 2 weeks ago. I didn't expect to be out of date already. I'll give it a try
